### ISSUE TITLE
Can't return array in component view

### ISSUE THREAD

**Issue by pobri19** [lhorie/mithril.js#869]
I want to be able to return an array like the below in my component so that I can add extra stuff to the div that encapsulates it, but without having to include that extra stuff inside the view. Returning it as a m("div", []) works but I don't want an unecessary div there for no reason. The console errors with this if I try and return an array:

```
view: function(ctrl) {
		return [
			m("input#username", {onchange: m.withAttr("value", ctrl.login.username), value: ctrl.login.username(), type: "text"}),
			m("input#password", {onchange: m.withAttr("value", ctrl.login.password), value: ctrl.login.password(), type: "password"}),
			m("button#login", {onclick: ctrl.submit}, "LOG IN")
		]
	}
```

```
Error: Component template must return a virtual element, not an array, string, etc.
...,fb.push(ib)}if(!l.tag&&fb.length)throw new Error("Component template must retur...
```

My use case is I'm trying to have those 3 inputs above belong to an authentication component, but I don't want my 4th input (the register button) to be contained within the authentication component, but I still want it inside the headerauth div like so:

```
	var HeaderAuth = function() {
		return m(".headerauth", [m.component(Auth), m("button#register", "REGISTER")])
	}
```

Here's someone else with a similar issue: https://github.com/mrsweaters/mithril-rails/issues/8

**Comment by Bondifrench** [comment]
You should put your inputs and button into a form.
Then instead of repeating`onchange` callbacks for every inputs you can use asymmetrical data bindings and the browser's event delegation system as explained by Leo [here](http://lhorie.github.io/mithril-blog/asymmetrical-data-bindings.html).

**Comment by pobri19** [comment]
Thanks @Bondifrench -- but the example I've got above has been updated now (originally I posted this to the wrong tracker, it looks like the article you cited now).

Under the caveats [here](https://lhorie.github.io/mithril/mithril.component.html#limitations-and-caveats), in particular number 1, says: "Nested component views must return either a virtual element or another component. Returning an array, a string, a number, boolean, falsy value, etc will result in an error."

My issue is that I want to be able to return an array of components without creating a wrapper div that is not needed, for example, this is what my code looks like now:

```javascript
view: function(ctrl) {
  return m(".globalconfigform", [
    m.component(GlobalConfigForm, {onsave: ctrl.save}),
    m.component(GlobalConfigList, {configs: ctrl.configs})
  ]);
}
```

Is there any way to do this without inserting an additional DOM element?

**Comment by tivac** [comment]
http://mithril.js.org/mithril.component.html#limitations-and-caveats

> Limitations and caveats
>
> ...
>
> 1. Nested component views must return either a virtual element or another component. Returning an array, a string, a number, boolean, falsy value, etc will result in an error.

**Comment by pobri19** [comment]
@tivac, thanks for the link I just posted.

**Comment by pobri19** [comment]
Yeah, I just posted that link. I'm just wondering if there is a workaround so that I don't have to insert an additional DOM element, other than refactoring the entire structure of my project because I rely on nested components.

**Comment by Bondifrench** [comment]
To my understanding, you can return an array of components in your top/parent view but a component in itself can not return an array.

**Comment by lhorie** [comment]
You can return arrays from simple functions (although in that case, you can't have state easily associated with the group). From components, returning arrays is currently not supported (but I *would* like to support it in the next major release)

**Comment by barneycarroll** [comment]
Sadly the error message isn't just there to annoy us. When components are invoked in the view, Mithril uses internal logic to determine whether the component should retrieve an existing controller or instantiate a new one. When a new one is instantiated (the default), it associates that controller instance with the root element of the corresponding view. Therefore, automatic nested component invocation relies on having a single DOM root element.

The alternative to this - the only way to invoke nested components before version 0.2 - is to take control of that logic yourself: manually instantiate the nested controller in the parent controller, bind it to a property, then pass it to the nested component view in the parent view instead of using automatic component invocation:

```
const Header = {
	controller(){
		this.auth = new( 
			Auth.controller || Object
		)
	},
	
	view : ctrl =>
		m( '.headerauth',
			Auth.view( ctrl.auth ),
			m( 'button#register',
				'REGISTER'
			)
		)
}
```

I found that I needed to systematise component invocation before 0.2 came along, and wrote a plugin for the purpose called [Modulator](https://gist.github.com/barneycarroll/0e3c1b9811b47c012b13) (back then 'components' were called 'modules'). The syntax is much uglier than Mithril's, but it's more flexible and allows multi-root nested components. If you find yourself requiring this pattern on a regular basis you might consider using Modulator instead of `m.component`.

All in all, I think the pragmatic approach is just to accept you will need root elements for components, even if they serve no purpose in the rendered DOM. AFAIK every other virtual DOM framework has the same requirement.
